### ISSUE TITLE
support global config to override defaults

### ISSUE THREAD

**Issue by casperdcl** [tqdm/tqdm#614]
1. look for e.g. "$HOME/.tqdm.config" to override defaults
2. alternatively/additionally use os.environ.get("TQDM_CONFIG", '')

helps with e.g. #370, #612, #619, #950, #1061, #1318

The second (os.environ) option would be great to fix https://github.com/tqdm/tqdm/issues/370#issuecomment-421809342 (have `tmux` alter `TQDM_CONFIG` to have `dynamic_ncols`)

**Comment by chengs** [comment]
For #612, setting something like DISABLE_TQDM in ~/.tqdm.config may be too dangerous. A user may forget this setting...

**Comment by ahundt** [comment]
I agree, this would be particularly useful for `dynamic_ncols`.

**Comment by ZeerakW** [comment]
There doesn't seem to have been much movement on this. I would tend to disagree with @chengs, I think if things are set from a tqdm.conf which needs to be manually created, people will look in that to identify the issue. In any case, having a global solution as in #950 would be much easier than locally adding the same disabling code to all TQDM loops in ones code.

**Comment by acroz** [comment]
I would also very much like global configuration from either environment variables or a configuration file, but for the `mininterval` parameter.

I'd be happy to put together a short proposal of how either solution might be added to the library for discussion. @casperdcl is there willingness from yourself or other maintainers to include something like this in the library?

**Comment by liam-ly** [comment]
I was looking for a context manager solution for #619 because I wanted to disable tqdm when using `pytest -s`. I came up with the answer below in case it helps anyone else. The idea is to monkeypatch the `__init__` function of the tqdm classes and then revert when leaving the context. I am using python 3.9.

```python
@contextlib.contextmanager
def monkeypatched(obj, name, patch):
    """Temporarily monkeypatch."""
    old_attr = getattr(obj, name)
    setattr(obj, name, patch(old_attr))
    try:
        yield
    finally:
        setattr(obj, name, old_attr)


@contextlib.contextmanager
def disable_tqdm():
    """Context manager to disable tqdm."""

    def _patch(old_init):
        return partialmethod(old_init, disable=True)

    with (
        monkeypatched(tqdm.std.tqdm, "__init__", _patch),
        monkeypatched(tqdm.asyncio.tqdm, "__init__", _patch),
        monkeypatched(tqdm.notebook.tqdm, "__init__", _patch),
    ):
        yield
```
In pytests, I can now use this as a decorator:
```python
from tqdm.auto import tqdm

@disable_tqdm()
def test_tqdm():
    assert tqdm().disable
```
or as a scoped fixture:
```python
@pytest.fixture(name="disable_tqdm", autouse=True, scope="session")
def disable_tqdm_fixture():
    """Disable tqdm in the test suite."""
    with disable_tqdm():
        yield

def test_tqdm():
    assert tqdm().disable
```

**Comment by mrclary** [comment]
I'd like to see some global control of tqdm as well, although I'd like to retain a fine-grained control as well. One workaround that I've come up with is to use a dummy logger to enable/disable tqdm instances.
```python
# __init__.py
import logging
_log = logging.getLogger('tqdm')
_log.setLevel(20)  # default level 20 will enable tqdm.
```
```python
# module.py
import time
from tqdm import tqdm
from . import _log

log = _log.getChild('module')

for i in tqdm(range(10), disable=not log.isEnabledFor(20)):
    time.sleep(1)
```

In this way, the global and hierarchal nature of the logger is available to the user to enable/disable a single progress bar or entire groups of progress bars. For example, all tqdm objects instantiated in this way can be disabled from anywhere with
```python
import logging
logging.getLogger('tqdm').setLevel(30)
```
Or, just a specific logger can be disabled with
```python
import logging
logging.getLogger('tqdm.module').setLevel(30)
```
Of course, this requires ownership of the source code, so it does not work with tqdm objects instantiated in third-party packages. However, I think it could be incorporated into the tqdm source code as a new feature and could also be combined with environment variables.

**Comment by Gilthans** [comment]
If you're still struggling with this, a workaround would be to create your own 'tqdm' module like this:
```
# tqdm.py
from tqdm.auto import tqdm as original_tqdm

tqdm = original_tqdm
tqdm.__init__ = functools.partialmethod(tqdm.__init__, mininterval=15, maxinterval=60, delay=15)
```
And then instead of importing the 'original' tqdm, you could import yours.
Obviously, you could also set this up to draw the values from environment variables.

**Comment by sisrfeng** [comment]
in `~/anaconda3/envs/t_pool3/lib/python3.9/site-packages/tqdm/std.py`, line 848, 
change `ascii=None, disable=True, unit='it', unit_scale=False,` to
             `ascii=None, disable=False, unit='it', unit_scale=False`

**Comment by technillogue** [comment]
+1, having an environment variable to disable tqdm would be great - I get many logs from upstream code using tqdm that's inconvenient to modify

**Comment by nikhilweee** [comment]
Found this on another issue (https://github.com/tqdm/tqdm/issues/370#issuecomment-556047962, thanks @casperdcl). Does this help?
```py
from functools import partial
from tqdm import tqdm as std_tqdm
tqdm = partial(std_tqdm, dynamic_ncols=True)
```

**Comment by casperdcl** [comment]
Thanks everyone for your patience! Just released [`tqdm==4.66.0`](https://github.com/tqdm/tqdm/releases/tag/v4.66.0) which includes support for env var overrides. I'm using `TQDM_MININTERVAL=5` in most of my CI jobs now :)

**Comment by adivekar-utexas** [comment]
The above did not work for me, so I have resorted to temporarily disabling stderr and reinstating it:

```
@contextmanager
def ignore_stderr():
    devnull = open(os.devnull, "w")
    stderr = sys.stderr
    sys.stderr = devnull
    try:
        yield
    finally:
        sys.stderr = stderr

with ignore_stderr():
    ## Call code that calls tqdm
```

This is obviously not ideal (warnings are also ignored) but it works
